heron/packing/src/java/org/apache/heron/packing/binpacking/TwoDimensionVectorBinPacking.java

```diff
@@ -0,0 +1,379 @@
+   */
+  private PackingPlanBuilder get2DVectorBinAllocation(PackingPlanBuilder packingPlanBuilder,
+                                                      SortingStrategy sortingStrategy,
+                                                      PackingPlan currentPackingPlan,
+                                                      Map<String, Integer> componentChanges)
+      throws ConstraintViolationException {
+    this.numContainers = currentPackingPlan.getContainers().size();
+
+    Map<String, Integer> componentsToScaleDown =
+        PackingUtils.getComponentsToScale(componentChanges, PackingUtils.ScalingDirection.DOWN);
+    Map<String, Integer> componentsToScaleUp =
+        PackingUtils.getComponentsToScale(componentChanges, PackingUtils.ScalingDirection.UP);
+
+    if (!componentsToScaleDown.isEmpty()) {
+      removeInstancesFromContainers(packingPlanBuilder, componentsToScaleDown, sortingStrategy);
+    }
+
+    if (!componentsToScaleUp.isEmpty()) {
+      assignInstancesToContainers(packingPlanBuilder, componentsToScaleUp, sortingStrategy);
+    }
+
+    return packingPlanBuilder;
+  }
+
+  /**
+   * Assigns instances to containers
+   *
+   * @param planBuilder existing packing plan
+   * @param parallelismMap component parallelism
+   */
+  private void assignInstancesToContainers(PackingPlanBuilder planBuilder,
+                                           Map<String, Integer> parallelismMap,
+                                           SortingStrategy sortingStrategy)
+      throws ConstraintViolationException {
+    List<ResourceRequirement> resourceRequirements
+        = getSortedInstances(parallelismMap.keySet(), sortingStrategy);
+    for (ResourceRequirement resourceRequirement : resourceRequirements) {
+      String componentName = resourceRequirement.getComponentName();
+      int numInstance = parallelismMap.get(componentName);
+      for (int j = 0; j < numInstance; j++) {
+        place2DVectorBinInstance(planBuilder, componentName);
+      }
+    }
+  }
+
+  /**
+   * Removes instances from containers during scaling down
+   *
+   * @param packingPlanBuilder existing packing plan
+   * @param componentsToScaleDown scale down factor for the components.
+   */
+  private void removeInstancesFromContainers(PackingPlanBuilder packingPlanBuilder,
+                                             Map<String, Integer> componentsToScaleDown,
+                                             SortingStrategy sortingStrategy) {
+
+    List<ResourceRequirement> resourceRequirements =
+        getSortedInstances(componentsToScaleDown.keySet(), sortingStrategy);
+
+    InstanceCountScorer instanceCountScorer = new InstanceCountScorer();
+    ContainerIdScorer containerIdScorer = new ContainerIdScorer(false);
+
+    for (ResourceRequirement resourceRequirement : resourceRequirements) {
+      String componentName = resourceRequirement.getComponentName();
+      int numInstancesToRemove = -componentsToScaleDown.get(componentName);
```

Comment:
componentsToScaleDown keeps negative counts? It feels a bit intuitive this way to me.